src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java

```diff
@@ -316,4 +694,223 @@ private AIState startBuilding()
+        {
+            return true;
+        }
+
+        @Nullable final ItemStack stack = BlockUtils.getItemStackFromBlockState(blockState);
+        if (stack == null)
+        {
+            Log.getLogger().error("Block causes NPE: " + blockState.getBlock());
+            return false;
+        }
+
+        final int slot = worker.findFirstSlotInInventoryWith(stack.getItem());
+        if (slot != -1)
+        {
+            getInventory().decrStackSize(slot, 1);
+            reduceNeededResources(block);
+        }
+        return true;
+    }
+
+    /**
+     * Reduces the needed resources by 1.
+     *
+     * @param block the block which has been used now.
+     */
+    public void reduceNeededResources(final Block block)
+    {
+        final AbstractBuilding workerBuilding = this.getOwnBuilding();
+        if (workerBuilding instanceof BuildingBuilder)
+        {
+            ((BuildingBuilder) workerBuilding).reduceNeededResource(block, 1);
+        }
+    }
+
+    private void spawnEntity(@Nullable final Entity entity)
+    {
+        if (entity != null && job instanceof JobBuilder)
+        {
+            final BlockPos pos = ((JobBuilder) job).getStructure().getOffsetPosition();
+
+            if (entity instanceof EntityHanging)
+            {
+                @NotNull final EntityHanging entityHanging = (EntityHanging) entity;
+
+                entityHanging.posX += pos.getX();
+                entityHanging.posY += pos.getY();
+                entityHanging.posZ += pos.getZ();
+                //also sets position based on tile
+                entityHanging.setPosition(
+                        entityHanging.getHangingPosition().getX(),
+                        entityHanging.getHangingPosition().getY(),
+                        entityHanging.getHangingPosition().getZ());
+
+                entityHanging.setWorld(world);
+                entityHanging.dimension = world.provider.getDimension();
+
+                world.spawnEntityInWorld(entityHanging);
+            }
+            else if (entity instanceof EntityMinecart)
+            {
+                @Nullable final EntityMinecart minecart = (EntityMinecart) entity;
+                minecart.posX += pos.getX();
+                minecart.posY += pos.getY();
+                minecart.posZ += pos.getZ();
+
+                minecart.setWorld(world);
+                minecart.dimension = world.provider.getDimension();
+
+                world.spawnEntityInWorld(minecart);
+            }
+        }
+    }
+
+    /**
+     * Create a save mining position for the miner.
+     * @param blockToMine block which should be mined or placed.
+     * @return the save position.
+     */
+    private BlockPos getNodeMiningPosition(BlockPos blockToMine)
+    {
+        if(((BuildingMiner) getOwnBuilding()).getCurrentLevel() == null)
+        {
+            return blockToMine;
+        }
+        Point2D pos = ((BuildingMiner) getOwnBuilding()).getCurrentLevel().getRandomNode().getParent();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'pos' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)